src/openapi/streaming/subscription-queue.js

```diff
@@ -29,38 +29,67 @@ function SubscriptionQueue(maxSize) {
  * - duplicates.
  * - ACTION_SUBSCRIBE action before ACTION_UNSUBSCRIBE.
  *
- * @param {Number} action - action to add to the queue.
+ * @param {Object} queuedAction - action with arguments to add to the queue.
  */
-SubscriptionQueue.prototype.enqueue = function(action) {
-    this.items.push(action);
+SubscriptionQueue.prototype.enqueue = function(queuedAction) {
 
-    if (this.items.length < this.maxSize) {
+    if (!queuedAction.action) {
+        throw new Error('Subscription queued action is invalid');
+    }
+    const { action } = queuedAction;
+
+    if (!this.canEnqueueAction(action)) {
         return;
     }
 
-    for (let i = this.items.length - 2; i >= 0; i--) {
-        const task = this.items[i];
-        const nextTask = this.items[i + 1];
-
-        if (task === ACTION_MODIFY_SUBSCRIBE && nextTask === ACTION_SUBSCRIBE) {
-            // Removing subscribe in such case to keep requested subscription modification.
-            this.items.splice(i + 1, 1);
-        } else if (task === ACTION_SUBSCRIBE && nextTask === ACTION_UNSUBSCRIBE ||
-            task === ACTION_UNSUBSCRIBE && nextTask === ACTION_SUBSCRIBE ||
-            task === ACTION_SUBSCRIBE && nextTask === ACTION_MODIFY_SUBSCRIBE ||
-            task === ACTION_MODIFY_SUBSCRIBE && nextTask === ACTION_UNSUBSCRIBE ||
-            task === nextTask
-        ) {
-            this.items.splice(i, 1);
-        }
+    if (action === ACTION_UNSUBSCRIBE ||
+        action === ACTION_SUBSCRIBE) {
+        this.reset();

```

Comment:
if there is a modify in the queue (say we are waiting for subscribe to happen) and we get a subscribe (or an unsubscribe then a subscribe), won't this forget the fact we need to do the modify?